### ISSUE TITLE
[1.12.4] docker network ls fails with EOF (panic)

### ISSUE THREAD

**Issue by DaneEveritt** [docker/docker#29411]
**Description**
After upgrading from `1.12.3` to `1.12.4` my application began crashing across multiple client's machines. After some quick investigation, it appears that listing networks using the API is returning an EOF causing errors to be thrown.

I tried to cURL the API endpoint, and confirmed that this was returning an empty reply. Additionally, other endpoints work correctly.

```
dane@daemon:~$ curl --unix-socket /var/run/docker.sock http://localhost/networks
curl: (52) Empty reply from server
```

**Steps to reproduce the issue:**
1. Potentially related, have a custom network defined.
1. Upgrade from `1.12.3` to `1.12.4`, attempt to `docker network ls`

**Describe the results you received:**
Received an error response from cURL

**Describe the results you expected:**
Should have received a listing of networks.

**Additional information you deem important (e.g. issue happens only occasionally):**
I do have a custom network defined, but I am unsure if that is the sole cause, or if there is something else going on. I'll report back when I can get a server to check that against.
**Output of `docker version`:**

```
Client:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Tue Dec 13 00:08:34 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Tue Dec 13 00:08:34 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 6
 Running: 0
 Paused: 0
 Stopped: 6
Images: 18
Server Version: 1.12.4
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 98
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge null host
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-47-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.842 GiB
Name: daemon
ID: ODHD:K37G:TWJV:7TSZ:O4YJ:674Y:XWQ3:O3IE:JSMC:QJII:IEO3:24X5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Virtual machine deployed using ESXi. Appears to be happening to clients using physical machines as well.

```
Linux daemon 4.4.0-47-generic #68-Ubuntu SMP Wed Oct 26 19:39:52 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.1 LTS"
```

**Comment by thaJeztah** [comment]
Could you check the daemon logs if there's any useful information there? 

Wondering if this is related to https://github.com/docker/docker/issues/28931#issuecomment-263882891

/cc @sanimej

**Comment by DaneEveritt** [comment]
```
Dec 14 10:47:00 daemon dockerd[80490]: time="2016-12-14T10:47:00.450169006-05:00" level=info msg="Loading containers: start."
Dec 14 10:47:00 daemon dockerd[80490]: ......time="2016-12-14T10:47:00.455132375-05:00" level=info msg="Firewalld running: false"
Dec 14 10:47:00 daemon dockerd[80490]: time="2016-12-14T10:47:00.513895899-05:00" level=warning msg="Failed to allocate ipam pool(s) for network \"pterodactyl_nw\" (1b0d1d353b3294a8e9daf6d6ca331850e4dfe82f86f280865afd6cd129fecd72): failed to parse pool request for address space \"LocalDefault\" pool \"\" subpool \"\": could not find an available, non-overlapping IPv6 address pool among the defaults to assign to the network"
Dec 14 10:47:00 daemon dockerd[80490]: time="2016-12-14T10:47:00.532215374-05:00" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
Dec 14 10:47:00 daemon dockerd[80490]: time="2016-12-14T10:47:00.555314426-05:00" level=info msg="Loading containers: done."
Dec 14 10:47:00 daemon dockerd[80490]: time="2016-12-14T10:47:00.555558166-05:00" level=info msg="Daemon has completed initialization"
Dec 14 10:47:00 daemon dockerd[80490]: time="2016-12-14T10:47:00.555675585-05:00" level=info msg="Docker daemon" commit=1564f02 graphdriver=aufs version=1.12.4
Dec 14 10:47:00 daemon systemd[1]: Started Docker Application Container Engine.
Dec 14 10:47:00 daemon dockerd[80490]: time="2016-12-14T10:47:00.567392604-05:00" level=info msg="API listen on /var/run/docker.sock"
Dec 14 10:47:03 daemon dockerd[80490]: http: panic serving @: runtime error: invalid memory address or nil pointer dereference
                                       goroutine 161 [running]:
                                       net/http.(*conn).serve.func1(0xc8203d8280)
                                               /usr/local/go/src/net/http/server.go:1389 +0xc1
                                       panic(0x1aa2760, 0xc82000e090)
                                               /usr/local/go/src/runtime/panic.go:443 +0x4e9
                                       net.networkNumberAndMask(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
                                               /usr/local/go/src/net/ip.go:433 +0x42
                                       net.(*IPNet).String(0x0, 0x0, 0x0)
                                               /usr/local/go/src/net/ip.go:483 +0x42
                                       github.com/docker/docker/api/server/router/network.buildIpamResources(0xc82031b900, 0x7f32d87f9e38, 0xc820412a00)
                                               /usr/src/docker/.gopath/src/github.com/docker/docker/api/server/router/network/network_routes.go:268 +0x4ec
                                       github.com/docker/docker/api/server/router/network.(*networkRouter).buildNetworkResource(0xc820795fb0, 0x7f32d87f0e70, 0xc820412a00, 0x40)
                                               /usr/src/docker/.gopath/src/github.com/docker/docker/api/server/router/network/network_routes.go:199 +0x3e8
                                       github.com/docker/docker/api/server/router/network.(*networkRouter).getNetworksList(0xc820795fb0, 0x7f32d88338b0, 0xc820794000, 0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000, 0xc8207a7e30, 0x0, 0x0)
                                               /usr/src/docker/.gopath/src/github.com/docker/docker/api/server/router/network/network_routes.go:46 +0x636
                                       github.com/docker/docker/api/server/router/network.(*networkRouter).(github.com/docker/docker/api/server/router/network.getNetworksList)-fm(0x7f32d88338b0, 0xc820794000, 0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000, 0xc8207a7e30, 0x0, 0x0)
                                               /usr/src/docker/.gopath/src/github.com/docker/docker/api/server/router/network/network.go:33 +0x74
                                       github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1(0x7f32d88338b0, 0xc820794000, 0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000, 0xc8207a7e30, 0x0, 0x0)
                                               /usr/src/docker/.gopath/src/github.com/docker/docker/api/server/middleware/version.go:56 +0x9aa
                                       github.com/docker/docker/api/server/middleware.UserAgentMiddleware.WrapHandler.func1(0x7f32d88338b0, 0xc8207a7ef0, 0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000, 0xc8207a7e30, 0x0, 0x0)
                                               /usr/src/docker/.gopath/src/github.com/docker/docker/api/server/middleware/user_agent.go:45 +0x68e
                                       github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000)
                                               /usr/src/docker/.gopath/src/github.com/docker/docker/api/server/server.go:139 +0x138
                                       net/http.HandlerFunc.ServeHTTP(0xc82029ac60, 0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000)
                                               /usr/local/go/src/net/http/server.go:1618 +0x3a
                                       github.com/gorilla/mux.(*Router).ServeHTTP(0xc82050b7c0, 0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000)
                                               /usr/src/docker/vendor/src/github.com/gorilla/mux/mux.go:98 +0x29e
                                       github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc820814170, 0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000)
                                               /usr/src/docker/.gopath/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x72
                                       net/http.serverHandler.ServeHTTP(0xc8203d8080, 0x7f32d87f5d08, 0xc8204dc000, 0xc82084e000)
                                               /usr/local/go/src/net/http/server.go:2081 +0x19e
                                       net/http.(*conn).serve(0xc8203d8280)
                                               /usr/local/go/src/net/http/server.go:1472 +0xf2e
                                       created by net/http.(*Server).Serve
                                               /usr/local/go/src/net/http/server.go:2137 +0x44e
```

**Comment by DaneEveritt** [comment]
I have fixed this by specifying the IPv6 gateway when making the network. I can close this if you'd like, or leave it open depending on if that EOF is intended or not.

**Comment by thaJeztah** [comment]
Thanks for the extra info @DaneEveritt!

I'll leave this open for @sanimej to have a look first

**Comment by mavenugo** [comment]
@aboch can u PTAL ?

**Comment by mavenugo** [comment]
This could be related to the fact that an existing network with `--ipv6` flag without an `subnet` is not properly handled during the upgrade (which requires the `--subnet`).
Its a corner case scenario and we might be okay with this. But lets have @aboch look at it to confirm

**Comment by vieux** [comment]
fixed by #29411
